### PR TITLE
Separate system threading library into a separate target

### PR DESCRIPTION
**Pull request by kintel**
This is just build system cleanup to avoid hardcoding a `-lpthread` linker option globally.

### CODE REVIEW

**Comment by benvanik** [comment]
Weird that only the big build is failing - and I'd have expected LLVM to set its dependencies right such that whatever we did wouldn't matter:
```
ERROR: /home/kbuilder/.cache/bazel/_bazel_kbuilder/1900d0fac5123d725c8d2e08b3f8c209/external/iree/iree/tools/BUILD:294:10: Linking of rule '@iree//iree/tools:iree-tblgen' failed (Exit 1): clang failed: error executing command /usr/bin/clang @bazel-out/k8-opt-exec-50AE0418/bin/external/iree/iree/tools/iree-tblgen-2.params

Use --sandbox_debug to see verbose messages from the sandbox clang failed: error executing command /usr/bin/clang @bazel-out/k8-opt-exec-50AE0418/bin/external/iree/iree/tools/iree-tblgen-2.params

Use --sandbox_debug to see verbose messages from the sandbox
bazel-out/k8-opt-exec-50AE0418/bin/external/llvm-project/llvm/_objs/Support/Process.o:Process.cpp:function llvm::sys::Process::SafelyCloseFileDescriptor(int): error: undefined reference to 'pthread_sigmask'
bazel-out/k8-opt-exec-50AE0418/bin/external/llvm-project/llvm/_objs/Support/Process.o:Process.cpp:function llvm::sys::Process::SafelyCloseFileDescriptor(int): error: undefined reference to 'pthread_sigmask'
```

Maybe try adding `//build_tools:threads` to `iree/tools/BUILD` iree-tblgen?

**Comment by benvanik** [comment]
Actually that's probably not what we want because iree-tblgen always needs threads regardless of what the runtime is being compiled for. We may need a separate target that is not conditional on the IREE_ENABLE_THREADS flag that is host/compiler-specific.

**Comment by benvanik** [comment]
(historically, due to coming from google3-land, we've conflated compiler and runtime flags - now that I've finally ripped out abseil we could probably start specifying for each of these what it's intended for - like //build_tools:runtime_threads vs. :compiler_threads)

**Review by ScottTodd — APPROVED** [review]

File: iree/tools/BUILD [hunk 1]
```diff
@@ -300,6 +300,7 @@ cc_binary(
     tags = ["hostonly"],
     deps = [
         "//build_tools:default_linkopts",
+        "//build_tools:threads",
```
Comment (ScottTodd) [inline iree/tools/BUILD:303]: We _could_ add a TODO here to remove this and unify with CMake once TF uses the same BUILD file for LLVM as we do.
Targets here in `tools/` tend to grow all sorts of deps though, so we can just look at that as part of general cleanup once in a while.

**Comment by ScottTodd** [comment]
> Actually that's probably not what we want because iree-tblgen always needs threads regardless of what the runtime is being compiled for. We may need a separate target that is not conditional on the IREE_ENABLE_THREADS flag that is host/compiler-specific.

Did you still have concerns with this?

**Comment by benvanik** [comment]
> > Actually that's probably not what we want because iree-tblgen always needs threads regardless of what the runtime is being compiled for. We may need a separate target that is not conditional on the IREE_ENABLE_THREADS flag that is host/compiler-specific.
> 
> Did you still have concerns with this?

yes - the compiler tools should not be based on IREE_ENABLE_THREADS, which is a runtime flag - then it would be a broken build to set IREE_ENABLE_THREADS=0 and try to build the compiler.

**Comment by benvanik** [comment]
> > > Actually that's probably not what we want because iree-tblgen always needs threads regardless of what the runtime is being compiled for. We may need a separate target that is not conditional on the IREE_ENABLE_THREADS flag that is host/compiler-specific.
> > 
> > 
> > Did you still have concerns with this?
> 
> yes - the compiler tools should not be based on IREE_ENABLE_THREADS, which is a runtime flag - then it would be a broken build to set IREE_ENABLE_THREADS=0 and try to build the compiler.

(we should rename it to IREE_RUNTIME_ENABLE_THREADS or something to make that clear)

**Comment by ScottTodd** [comment]
> > > Actually that's probably not what we want because iree-tblgen always needs threads regardless of what the runtime is being compiled for. We may need a separate target that is not conditional on the IREE_ENABLE_THREADS flag that is host/compiler-specific.
> > 
> > 
> > Did you still have concerns with this?
> 
> yes - the compiler tools should not be based on IREE_ENABLE_THREADS, which is a runtime flag - then it would be a broken build to set IREE_ENABLE_THREADS=0 and try to build the compiler.

kk, undo'ing my stamp :P

**Comment by kintel** [comment]
I'm not longer sure this is a problem worth solving. A better use of time could be to help Tensorflow update their Bazel build.
My latest commit tries to shoehorn conditional linking into the BUILD file, but I'm guessing there must be a better way of expression conditionals in Bazel.

PS. How do we tag CLs as "do not merge" ?

**Comment by ScottTodd** [comment]
> I'm not longer sure this is a problem worth solving. A better use of time could be to help Tensorflow update their Bazel build.
> My latest commit tries to shoehorn conditional linking into the BUILD file, but I'm guessing there must be a better way of expression conditionals in Bazel.
> 
> PS. How do we tag CLs as "do not merge" ?

convert to draft probably: https://github.blog/changelog/2020-04-08-convert-pull-request-to-draft/

(if you see this button, permissions are a little weird)

image

or edit the PR title, as you've done

**Pull request closed without merging**